\documentclass[a4,center,fleqn]{NAR}

\usepackage{todonotes}

% Enter dates of publication
\copyrightyear{2008}
\pubdate{31 July 2009}
\pubyear{2009}
\jvolume{37}
\jissue{12}

%\articlesubtype{This is the article type (optional)}

\begin{document}

\title{H. pylori Homology Database: why special databases are important in bacterial research}

\author{%
Markus Joppich\,$^{1,*}$,
Luisa Jimenez\,$^{3}$
and Ralf Zimmer\,$^1$%
\footnote{To whom correspondence should be addressed.
Tel: +49 89 2180 4045; Email: joppich@bio.ifi.lmu.de}}

\address{%
$^{1}$Affiliation of Corresponding Author
and
$^{2}$Affiliation of Both Co-Authors}
% Affiliation must include:
% Department name, institution name, full road and district address,
% state, Zip or postal code, country

\history{%
Received January 1, 2018;
Revised February 1, 2018;
Accepted March 1, 2018}

\maketitle

\begin{abstract}
	To study microorganisms it is necessary to evaluate phenotypes, which include biochemical and visual characteristic, and relation with its environment.
	Additionally the advance of sequencing technologies allow access to information about the organism's genome and how it is transcribed from multiple exemplars, that combined with its phenotype, results in a more complete picture of a microorganism.
	However, large quantities of genomes are being submitted every day at a pace challenging the capacities of analysis for those studying an organism.
	To date most searches for homology genes / proteins are done with the objective of find unknown function based on orthology or paraorthology to already known genes / proteins.
	This kind of searches are important to estimate the function of unknown proteins or genes. However many of the advantages of obtaining data from several clones from the same species are not being fully exploited at the moment, and if the researcher wants to take advantage of this homologies it need to rely on BLASTs or in a unanimous annotation of the genes of interested. The last is hardly achievable, leaving an tenuous and limited search through alignment.
	In order to show the advantages of genome / proteome information from several strains , we have developed a database using as model organism \textit {Helicobacter pylori}.
	With the use of the database we have found that strains from \textit{H. pylori} use tryptophan in proteins in a strain-specific way with potential changes in the function of membrane related and cation-binding proteins.
	We extended this analysis to other bacterial species adapting this database to their genome information, showing that it can be adapted to any microrganism of interest of which at least one complete genome is available in any genomic database together with several strains.
\end{abstract}


\section{Introduction}

The idea that similar protein or gene sequences have a higher probability of fulfilling the same or similar function has been the foundation for searches based on alignments, being BLAST the most popular. Although similarity could mean analogous function, the lack of similitude does not exclude it. Alignments can help to discover the function of unknown genes or proteins under the premise that the function of a homologous sequence is already known, as variations of in sequences are considered part of the evolutionary process (reviewed by Pearson WR 2013).\\
Until 2015 there were genome sequences from 50 bacterial and 11 archaeal phyla available with a total of around 14000 (February 2015, NCBI) (Land M 2015). Today, over three years later,  this number has increase to a total of 133148 (March 15, 2018. NCBI Genome). The rate at which genomes are being published increases the possibility of finding a homologous gene or protein. At the same time, analysis of multiple genomes from different specimens (strains) belonging to same species allow to estimate normal variations of the organisms in ecosystems. However, this data growth rate challenges the ability to analyze it in a coherent manner based on already available information.\\	 

One of the many difficulties in the analysis of homologous proteins within one species is the variation of annotations in databases. Each submitted genome uses a different annotation for the genes/ proteins found in their data. Some are achieved through automated alignment and homology assignment. Other annotations are based on historic references made by the researchers at submission. One case are proteins components of the Type IV secretion System (T4SS). Bacterial species that present components for a T4SS, like Bordetella pertussis, Helicobacter pylori and Legionella pneumophila, have had their components described in different ways depending on the researcher submitting genomes,  or as a result of an automated homology search made. One example of the consequences for these variation in annotation of genomes is the L. pneumophila's component DotL (From the Dot/Icm T4SS). It can be found as across literature and genomes as DotL, IcmO, or referred as VirD4 homologue based on its resemblance to the first T4SS defined in Agrobacterium tumefasciens . However, if the text search is done using VirD4 in L. pneumophila, none of these proteins will appear. Instead the results will include LvhD4, VirD4 component, conjugal transfer protein TraG, hypothetical protein or Type IV secretory system conjugative DNA transfer family protein. 
% **************************************************************
% Keep this command to avoid text of first page running into the
% first page footnotes
\enlargethispage{-65.1pt}
% **************************************************************

This kind of variations in annotations makes the work with bacterial proteins names extremely difficult and frustrating. This case is not isolated: all other bacteria named above present the same challenges while working with T4SS, and although the variations in annotations of the T4SS could have been an exception, sadly it is more the rule for many of the annotations of genes in their genome.\\
Although there are several databases for homology search (HomoloGene from NCBI (Ref), OrtholugeDB (ref), OrthoDB (ref)), none of them are adaptable and upgradable independently for one species of bacteria and complementary information emerging with new technologies as it is transcriptomes, small RNAs and primary and secondary promotors. 
In order to work with all homologous genes from one organism, the annotation's variations forces the use of sequence BLAST against the taxID of the species of interest, resulting in a list of proteins/genes with homologies to the query. This alignments can then be exported and used for analysis. Although functional this is not the most efficient way to obtain several homologous genes from all strains sequenced.\\The database presented here is designed to eliminate the need of BLAST of each protein against its own database, facilitate the comparison of genomic and proteomic data available in the EMBL genome database from different strains from the same organism, and it is adaptable to the addition of transcriptome data available the standard strain 26695. \\
To evaluate the effectiveness and usefulness of the database, we have chosen as model organism \textit{Helicobacter pylori}, an epsilon proteobacteria present in the human stomach and associated with gastric pathologies(ref). H. pylori is a natural competent bacterium, able to capture and integrate extracellular DNA into its genome. At the same time it present a high genetic variability between strains. The first strain sequenced from this organism was the strain 26695 (ref). Its open-reading frames (ORFs or Locus Tags) have been used for the definition of many  \textit{H. pylori} characteristic features, like the Cag Pathogenicity Island (Ref), the Outer Membrane Protein families (ref) and recently to describe the transcriptome of \textit{H. pylori}(Sharma et al). The strain J99 sequence followed and its locus tags have been used as synonym to describe proteins or genes. While these descriptions were sufficient to analyze single genes in the following years, the exponential growth of new genomes available and variations in annotations between strains makes it difficult to find the known homologue genes.
During experiments requiring semiquantitative analysis of western blot signals, the normalization of sample load for each bacterial strain in a polyacrylamide gel electrophoresis required the detection of proteins in a way compatible with further immunological analysis. The use of 2,2,2, trichloroethanol (TCE) with ultraviolet activation (ref) allowed the estimation of protein present in the gel and the use of  same samples for western blot analysis. The TCE allows fluorescent detection of tryptophan present in proteins at 305 nm without creating a crosslink of proteins with the acrylamide gel. Imaging of tryptophan revealed that different strains presented variation in their patterns showing changes in tryptophan content in their proteins (Rojas et al, submitted). For this analysis, we were confronted with the need to identify the right homologous proteins in both genomes for quantification of tryptophan residues. This proved tedious and not really feasible if more strains would be analyzed. Therefore we created a database that allows the identification of homologous groups of proteins across different genomes independently of their annotations, and allows the introduction of transcriptome information if available for the organism. To show the biological reference using the database we show here that not only H. pylori uses tryptophan in an specific way, but ..... and ... as well, while other bacterial species like \textit{Escherichia coli} and \textit{Campylobacter pylori} restrict the variations in tryptophan residues.
Text \cite{2,3}.

Text \cite{4}.


\section{MATERIALS AND METHODS}

\subsection{Current Homology Databases}

There exist several homology databases.


\subsection{Helicobacter Homology Database}


\subsubsection{Finding Homologies}

\paragraph{Homology Groups (HOMID)} are simple 1:1 homologies.
\paragraph{Combined Homology Groups (COMBID)} are simple 1:n homologies.
\paragraph{Homology Groups (MULCOM)} are simple m:n homologies.

\subsubsection{Integrating Cross-References}

Uniprot and Cyntia

\subsubsection{Improving Pfam Results}

How many Pfams did I find, how many have been reported by Uniprot

\subsubsection{Improving Gene Ontology relations}

\subsubsection{Integrating SwissModel Protein Projections}



\section{RESULTS}

We have created the Helicobacter Homology Database.

\subsection{Found Homologies}

\begin{figure}
	\missingfigure{Distribution of homology sizes}
	\caption{Distribution of homology group sizes}
	\label{fig:hom_sizes}
\end{figure}


\begin{table}[b]
\tableparts{%
\caption{Found homologies in the core genomes.}
\label{table:01}%
}{%
\begin{tabular*}{\columnwidth}{@{}lllll@{}}
\toprule
Strain & Proteins/Genes & Regular & Combined & Multi-Combined
\\
& & Homologies & Homologies & Homologies
\\
\colrule
Row 1 & Row 1 & Row 1 & -- & --
\\
Row 2 & Row 2 & Row 2 & Row 2 & Row 2
\\
\botrule
\end{tabular*}%
}
{The core genomes consist of ...}
\end{table}

\subsection{Using textmining to improve Gene Ontology}

We have applied the Pfam hmmer model to all proteins in our database to get accurate Pfam categories.
We could thereby find all annotated results in uniprot, and additional ones ....

Additionally we textmined the Pfam and Interpro descriptions with the GO keywords to receive new relations. We found bla existing relations and some more additional ones.
We did not report a new relation if our newly found relation is a child of an exisiting one.

We evaluated our tool on the set of interesting genes from the other paper.


\begin{table}[b]
	\tableparts{%
		\caption{Textmining results for InterPro and Pfam. Additionally the number of reported entries from Uniprot are given in brackets. The overlap between InterPro and Pfam is exact match, for Gene Ontology an overlap is given if a the same element, or a child is found.}
		\label{table:tm_performance}%
	}{%
		\begin{tabular*}{\columnwidth}{@{}lllll@{}}
			\toprule
			Col. head 1 & Col. head 2 & Col. head 3 & Col. head 4 & Col. head 5
			\\
			& (\%) & (s$^{-1}$) & (\%) & (s$^{-1}$)
			\\
			\colrule
			Row 1 & Row 1 & Row 1 & -- & --
			\\
			Row 2 & Row 2 & Row 2 & Row 2 & Row 2
			\\
			\botrule
		\end{tabular*}%
	}
	{The core genomes consist of ...}
\end{table}

\subsection{Meta-analysis of all H. pylori species}

We want to know, how are the clusters distributed?

All proteins (roughly 10000) form 25000 homology clusters?
I would have expected roughly 2000 clusters ...

\begin{figure}
	\missingfigure{MLST tree for given organisms}
	\caption{MLST tree for given 61 organisms.}
	\label{fig:mlst_tree}
\end{figure}

\begin{figure*}[t]
	\begin{center}
		\missingfigure{something}
	\end{center}
	\caption{
		Clustermap resulting from whether an organism is in a homology group or not.
	}
	\label{fig:hom_clustermap}
\end{figure*}

Also, could we find anything useful with Cyntia's data? \todo{Is there anything we can do with that?}

\section{DISCUSSION}

\subsection{Accessibility of Information}

Briefly discuss the user interface here.

\begin{figure}
	\missingfigure{Screenshot from User Interface}
	\caption{HPDB user interface.}
	\label{fig:db_ui}
\end{figure}

\subsection{A general database misses a lot}
We could see that there is a lot of missing information in general databases like uniprot.

First, updating such large databases is time consuming.
Second, things should work overall good, however, might break with little details.

\subsection{How accurate are the homology clusters?}

Check whether the organisms cluster into groups ...

... compare resulting clusters with MLST data?

\section{CONCLUSION}

Natural variations at the genome and protein level within a species can show us the flexibility of the genetic code in each organism. Although the search for homology across species is the most used method for search of  function, the variations withing a species can tell us about the intrinsic activity and tolerance levels of mutations for a single protein inside their species. Our Database allows to overcome the irregularities in gene annotations present in genome databases, and allows to adapt the emergent information about an microorganism like transcriptome and small RNAs, making it unique compared to other homologue search databases, where the primary function is to find the homology across different species.



\section{ACKNOWLEDGEMENTS}

We want to thank .... for ...., and .... for his/her support with ....


\subsubsection{Conflict of interest statement.} None declared.
\newpage


\begin{thebibliography}{4}

% Format for article
\bibitem{1}
Author,A.B. and Author,C. (1992)
Article title.
\textit{Abbreviated Journal Name}, \textbf{5}, 300--330.

% Format for book
\bibitem{2}
Author,D., Author,E.F. and Author,G. (1995)
\textit{Book Title}.
Publisher Name, Publisher Address.

% Format for chapter in book
\bibitem{3}
Author,H. and Author,I. (2005)
Chapter title.
In
Editor,A. and Editor,B. (eds),
\textit{Book Title},
Publisher Name, Publisher Address,
pp.\ 60--80.

% Another article
\bibitem{4}
Author,Y. and Author,Z. (2002)
Article title.
\textit{Abbreviated Journal Name}, \textbf{53}, 500--520.

\end{thebibliography}

\end{document}
